### PR TITLE
Add storage and resend for validators preferences

### PR DESCRIPTION
**Pull request by 0xpanoramix**
**Description:**

This PR adds a way to send validators preferences at a regular interval to the relays.
It follows the implementation proposed in #141 and is part of the steps to achieve v1 as described in #95.

**Problem(s) & goal(s):**

//

**Additional context & references:**

This PR adds a method to handle the logic of sending the preferences to the relays, which is used by both the standard handler and a new method used to send the already registered prefs at a provided interval.

Several questions still remains regarding the implementation, see #141 for more.

---

**I have run these commands:**

* [x] `make lint`
* [x] `make test`
* [x] `make run-mergemock-integration`
* [x] `go mod tidy`

### CODE REVIEW

**Comment by metachris** [comment]
Very nice, and thoughtful! I noticed that `registerValidatorAtInterval` is not yet used, but it's still marked as wip so I'm sure you're aware ^^

File: cmd/mev-boost/main.go [hunk 1]
```diff
@@ -21,16 +21,18 @@ var (
...
 	// defaults
-	defaultListenAddr         = getEnv("BOOST_LISTEN_ADDR", "localhost:18550")
-	defaultRelayTimeoutMs     = getEnvInt("RELAY_TIMEOUT_MS", 2000) // timeout for all the requests to the relay
-	defaultRelayCheck         = os.Getenv("RELAY_STARTUP_CHECK") != ""
-	defaultGenesisForkVersion = getEnv("GENESIS_FORK_VERSION", "")
+	defaultListenAddr                = getEnv("BOOST_LISTEN_ADDR", "localhost:18550")
+	defaultRelayTimeoutMs            = getEnvInt("RELAY_TIMEOUT_MS", 2000) // timeout for all the requests to the relay
+	defaultRelayCheck                = os.Getenv("RELAY_STARTUP_CHECK") != ""
+	defaultGenesisForkVersion        = getEnv("GENESIS_FORK_VERSION", "")
+	defaultRegisterValidatorInterval = getEnvInt("REGISTER_VALIDATOR_INTERVAL_SEC", 5)
```
Comment (Ruteri) [inline cmd/mev-boost/main.go:28]: Let's think about a sane default, if all of the validators send their registration every 5 seconds it will effectively be a DDoS attack on the unfortunate relay on the receiving end
Slot is 12 seconds, that's a sane lower bound
An epoch is 384 seconds, and validators are asked to re-register once per epoch so it's an upper bound
I think something like a fourth (96s) of an epoch is already frequent enough

File: server/service.go [hunk 2]
```diff
@@ -129,6 +146,60 @@ func (m *BoostService) handleStatus(w http.ResponseWriter, req *http.Request) {
...
+	log := m.log.WithField("method", "registerValidatorAtInterval")
+
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-done:
+			return
+		case <-ticker.C:
```
Comment (Ruteri) [inline server/service.go:196]: One issue that arises is that we can send twice in a short interval (if a registration request comes near the timer tick)
In this approach it's not easy to solve it, and I think we could come at this from a different angle:
* make registerValidatorAtInterval the only place responsible for sending the registration
* listen on both the timer, and a new channel that accepts new registration requests
* on new registration request, send the preferences and reset the timer

This has some nice benefits:
* m.vp is now an object local to registerValidatorAtInterval
* validator preferences struct can become just a plain old map
* no need for mutex
* after a registration you will correctly wait until the next interval

The only downside is that you'll need to return the registration status back over a channel (or a callback), but it should be okay.

We can submit a PR to this PR if you prefer not implementing it

File: cmd/mev-boost/main.go (hunk 1 above)
Comment (metachris) [reply cmd/mev-boost/main.go:28]: even resending once an epoch should be fine (and instantly on receiving)

File: server/service.go (hunk 2 above)
Comment (0xpanoramix) [reply server/service.go:196]: Totally agree with this comment ! Thanks a lot for raising the issue and proposing an alternative 👍
I can make the updates don't worry.

File: cmd/mev-boost/main.go (hunk 1 above)
Comment (0xpanoramix) [reply cmd/mev-boost/main.go:28]: I agree with your comments, but it seems like it's a bad idea to let the user decide of the interval in the end, wdyt ?

File: server/service.go (hunk 2 above)
Comment (0xpanoramix) [reply server/service.go:199]: Added in de04d2f038dcde4efcd2bc8b7928911a762db457

cc @Ruteri

File: server/service.go [hunk 3]
```diff
@@ -164,32 +237,10 @@ func (m *BoostService) handleRegisterValidator(w http.ResponseWriter, req *http.
...
-			numSuccessRequestsToRelay++
-		}(relay.Address)
-	}
-
-	// Wait for all requests to complete...
-	wg.Wait()
+	// Send the payload to the goroutine responsible for handling the resend at interval
+	m.newRegistrationsRequests <- payload
+	// Block until we get the number of successful requests back from this goroutine
+	numSuccessRequestsToRelay := <-m.numSuccessRequestsToRelay
```
Comment (Ruteri) [inline server/service.go:243]: Once context (timeout) handling is implemented this will become a race, the result channel (or callback) should be specific to the request.
I think pushing a pair of (payload, result chan) would be a clean way to solve it that will also work nicely with context timeouts.

File: server/service.go [hunk 4]
```diff
@@ -56,10 +57,17 @@ type BoostService struct {
 	serverTimeouts       HTTPServerTimeouts
 
 	httpClient http.Client
+
+	// Used to stop registerValidatorAtInterval
+	done chan bool
+	// Used by registerValidator to share new incoming registration request with the goroutine holding the ticker
```
Comment (Ruteri) [inline server/service.go:63]: It'd be great to avoid shared fields as they usually introduce a race

File: server/service.go [hunk 5]
```diff
@@ -56,10 +57,17 @@ type BoostService struct {
 	serverTimeouts       HTTPServerTimeouts
 
 	httpClient http.Client
+
+	// Used to stop registerValidatorAtInterval
+	done chan bool
```
Comment (Ruteri) [inline server/service.go:62]: Another way to implement shutdowns will be to have a context with cancel, and storing the cancel function.

File: server/service.go [hunk 6]
```diff
@@ -110,10 +122,15 @@ func (m *BoostService) StartHTTPServer() error {
 		IdleTimeout:       m.serverTimeouts.Idle,
 	}
 
+	// Start separate process to send validator preferences at regular interval.
+	go m.registerValidatorAtInterval(time.Second*384, m.done)
```
Comment (Ruteri) [inline server/service.go:126]: I'm guessing this should be registerValidatorIntervalSec

File: server/service.go [hunk 7]
```diff
@@ -384,3 +435,7 @@ func (m *BoostService) handleGetPayload(w http.ResponseWriter, req *http.Request
 		return
 	}
 }
+
+func (m *BoostService) shutdown() {
+	m.done <- true
```
Comment (metachris) [inline server/service.go:440]: using a context would be better because if there's multiple things to shut down, this channel would only work for one of them. could be okay for now, and we can also do this in another PR.

File: server/service.go [hunk 8]
```diff
@@ -129,6 +146,62 @@ func (m *BoostService) handleStatus(w http.ResponseWriter, req *http.Request) {
...
+
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-done:
+			// mev-boost has probably stopped
+			return
+		case payload = <-m.newRegistrationsRequests:
```
Comment (metachris) [inline server/service.go:194]: this will overwrite the previous payload, but it's conceivable that there's multiple validators using one BN that would then overwrite each others registrations. perhaps better to make this additive?

Comment (Ruteri) [reply server/service.go:194]: I'm actually not sure about it, the expectation is to have a single BN connecting to mev-boost, and forcing this to be additive would mean you cannot ever remove a registration.

File: server/service.go (hunk 5 above)
Comment (0xpanoramix) [reply server/service.go]: I agree, that's what they've been made after all.
Fixed in a2bb054510f72b4545958aa409ad48ffe7700382

File: server/service.go (hunk 6 above)
Comment (0xpanoramix) [reply server/service.go]: Missed that, thanks ! Fixed in e1d657247437af9b985129d2e6a1ec7f1b8891c6.

File: server/service.go (hunk 8 above)
Comment (metachris) [reply server/service.go]: @Ruteri expectation is a single BN, but this single BN could have _many_ validator clients connected! We can not ignore this case.

One approach could be to add a time to each registration and remove old ones.

File: server/service.go [hunk 9]
```diff
@@ -129,6 +147,62 @@ func (m *BoostService) handleStatus(w http.ResponseWriter, req *http.Request) {
...
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-m.ctx.Done():
+			// mev-boost has probably stopped
+			return
+		case srr := <-m.registrationsRequests:
+			// registerValidator has received new registrations and forwards them to here
+			srr.numSuccessRequestsToRelay <- m.sendValidatorPreferences(log, srr.preferences)
```
Comment (metachris) [inline server/service.go:197]: `payload` isn't updated

Comment (0xpanoramix) [reply server/service.go:197]: 👍 
Fixed in a3d5a4b578ce5c133f749112f71ba096230488f4

**Comment by metachris** [comment]
It was closed because the temporary develop branch was deleted after merging into main. Please point that PR towards the main branch, I seem to not be able to do it from my user, and can't reopen until then :)
